### PR TITLE
feat: Add `to_local_datetime` & `from_local_datetime`

### PR DESCRIPTION
**Pull request by ion-elgreco**
An attempt at allowing a date to be converted to local timezone by converting to NaiveDataTime and then setting the timezone to None.

In follow-up, I will rename the other naive_local_opt functions to make that part consistent and I will expressify the ambiguous parameter in `from_local_datetime` @MarcoGorelli 

- closes https://github.com/pola-rs/polars/issues/12761

### CODE REVIEW

File: crates/polars-ops/src/chunked_array/datetime/replace_time_zone.rs [hunk 1]
```diff
@@ -82,3 +84,55 @@ pub fn replace_time_zone(
...
+                        &parse_time_zone(convert_tz)?,
+                        ndt,
+                    )?,
+                ))
+            }),
+            _ => Ok(datetime.0.apply(|_| None)),
+        },
+        _ => try_binary_elementwise(datetime, convert_tz, |timestamp_opt, ambiguous_opt| match (
+            timestamp_opt,
+            ambiguous_opt,
```
Comment (MarcoGorelli) [inline crates/polars-ops/src/chunked_array/datetime/replace_time_zone.rs:121]: should this be `convert_tz_opt`?

File: py-polars/polars/expr/datetime.py [hunk 2]
```diff
@@ -1423,6 +1423,52 @@ def replace_time_zone(
             self._pyexpr.dt_replace_time_zone(time_zone, ambiguous._pyexpr)
         )
 
+    def convert_to_local_time_zone(
+        self,
+        local_time_zone: IntoExpr,
+    ) -> Expr:
+        """
+        Converts to local timezone and then replaces timezone with None.
```
Comment (MarcoGorelli) [inline py-polars/polars/expr/datetime.py:1431]: How about "Convert to local datetime in given time zone."?

File: py-polars/polars/expr/datetime.py [hunk 3]
```diff
@@ -1423,6 +1423,52 @@ def replace_time_zone(
...
+
+        Parameters
+        ----------
+        local_time_zone
```
Comment (MarcoGorelli) [inline py-polars/polars/expr/datetime.py:1435]: just `time_zone` should be enough

File: py-polars/polars/expr/datetime.py [hunk 4]
```diff
@@ -1423,6 +1423,52 @@ def replace_time_zone(
...
+        >>> from datetime import datetime
+        >>> df = pl.DataFrame(
+        ...     {
+        ...         "date_col": [datetime(2020, 10, 10)] * 3,
+        ...         "timezone": ["Europe/London", "Africa/Kigali", "America/New_York"],
+        ...     }
+        ... ).with_columns(pl.col("date_col").dt.replace_time_zone("UTC"))
+        >>> df.with_columns(
+        ...     pl.col("date_col")
+        ...     .dt.convert_to_local_time_zone("timezone")
```
Comment (MarcoGorelli) [inline py-polars/polars/expr/datetime.py:1454]: is this the API we want? how about `.dt.convert_to_local_time_zone(pl.col("timezone"))`, and if you pass a string, then that's the time zone?

File: py-polars/polars/expr/datetime.py [hunk 5]
```diff
@@ -1423,6 +1423,52 @@ def replace_time_zone(
             self._pyexpr.dt_replace_time_zone(time_zone, ambiguous._pyexpr)
         )
 
+    def convert_to_local_time_zone(
```
Comment (MarcoGorelli) [inline py-polars/polars/expr/datetime.py:1426]: not sure about this name - maybe just `to_naive_local`?

**Review by MarcoGorelli — COMMENTED** [review]
thanks for starting this, got some initial comments

File: py-polars/polars/expr/datetime.py (hunk 5 above)
Comment (ion-elgreco) [reply py-polars/polars/expr/datetime.py:1426]: I found this naive_local naming that I saw in chrono super confusing 😆

File: py-polars/polars/expr/datetime.py (hunk 4 above)
Comment (MarcoGorelli) [reply py-polars/polars/expr/datetime.py:1454]: hmm true, thanks

**Review by MarcoGorelli — COMMENTED** [review]
also not sure about names, will sleep on it - in the meantime, left a note on the implementation

File: crates/polars-arrow/src/legacy/kernels/time.rs [hunk 6]
```diff
@@ -70,3 +70,16 @@ pub fn convert_to_naive_local_opt(
...
     }
 }
+
+#[cfg(feature = "timezones")]
+pub fn convert_to_new_timezone_and_naive_local(
+    from_tz: &Tz,
+    to_tz: &Tz,
+    ndt: NaiveDateTime,
+) -> PolarsResult<NaiveDateTime> {
+    // ndt is already converted to UTC timezone
```
[the remark this answers is not in the document]
Comment (ion-elgreco) [reply crates/polars-arrow/src/legacy/kernels/time.rs:80]: Yeah indeed, the note was just for myself during coding

File: crates/polars-arrow/src/legacy/kernels/time.rs [hunk 7]
```diff
@@ -70,3 +70,16 @@ pub fn convert_to_naive_local_opt(
         LocalResult::None => None,
     }
 }
+
+#[cfg(feature = "timezones")]
+pub fn convert_to_new_timezone_and_naive_local(
+    from_tz: &Tz,
+    to_tz: &Tz,
+    ndt: NaiveDateTime,
+) -> PolarsResult<NaiveDateTime> {
```
[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply crates/polars-arrow/src/legacy/kernels/time.rs]: yeah when it can fail

File: crates/polars-arrow/src/legacy/kernels/time.rs [hunk 8]
```diff
@@ -70,3 +70,16 @@ pub fn convert_to_naive_local_opt(
         LocalResult::None => None,
     }
 }
+
+#[cfg(feature = "timezones")]
+pub fn convert_to_new_timezone_and_naive_local(
```
Comment (MarcoGorelli) [inline crates/polars-arrow/src/legacy/kernels/time.rs:75]: not sure about this name either

how about
- first, in a separate PR, you rename `convert_to_naive_local` to `replace_time_zone_in_naive_utc`. We're going from one naive utc representation to another, so maybe that's a more appropriate name (maybe also add a docstring with an example? optional)
- then, here, this one could be called `naive_utc_to_naive_local_in_new_time_zone`?

File: py-polars/polars/expr/datetime.py (hunk 5 above)
Comment (MarcoGorelli) [reply py-polars/polars/expr/datetime.py:1426]: True, but what does "local time zone" mean? A datetime can be local, not sure about this one

Maybe just `to_local_datetime`?

We'd also need this method to be added to docs/user-guide/transformations/time-series/timezones.md

File: crates/polars-arrow/src/legacy/kernels/time.rs (hunk 8 above)
Comment (MarcoGorelli) [reply crates/polars-arrow/src/legacy/kernels/time.rs:75]: one sec, before you do too much work here, just checking that there's no opposition to this feature

Comment (ion-elgreco) [reply crates/polars-arrow/src/legacy/kernels/time.rs:75]: Sure :)

File: py-polars/polars/expr/datetime.py [hunk 9]
```diff
@@ -1423,6 +1423,83 @@ def replace_time_zone(
...
+        │ ---                     ┆ ---              ┆ ---                 │
+        │ datetime[μs, UTC]       ┆ str              ┆ datetime[μs]        │
+        ╞═════════════════════════╪══════════════════╪═════════════════════╡
+        │ 2020-10-10 00:00:00 UTC ┆ Europe/London    ┆ 2020-10-10 01:00:00 │
+        │ 2020-10-10 00:00:00 UTC ┆ Africa/Kigali    ┆ 2020-10-10 02:00:00 │
+        │ 2020-10-10 00:00:00 UTC ┆ America/New_York ┆ 2020-10-09 20:00:00 │
+        └─────────────────────────┴──────────────────┴─────────────────────┘
+
+        """
+        local_tz = parse_as_expression(tz, str_as_lit=False)
```
Comment (MarcoGorelli) [inline py-polars/polars/expr/datetime.py:1467]: I don't think there's a general rule about whether a string should be parsed as literal or expression, but in other functions with time zones, the time zone is parsed as a literal - so for consistency, could we do that here too please?

`.dt.to_naive_local('Asia/Kathmandu')` is more ergonomic than `.dt.convert_time_zone('Asia/Kathmandu').dt.datetime()` so people might reasonably want to use this with just a single time zone

File: py-polars/polars/expr/datetime.py [hunk 10]
```diff
@@ -1423,6 +1423,83 @@ def replace_time_zone(
...
+        return wrap_expr(self._pyexpr.dt_to_naive_local(local_tz))
+
+    def from_naive_local(
+        self,
+        naive_tz: IntoExpr,
+        out_tz: str,
+        ambiguous: Ambiguous = "raise",
```
Comment (MarcoGorelli) [inline py-polars/polars/expr/datetime.py:1474]: ok for a follow-up, but this would ideally be an expression too, right?

File: py-polars/tests/unit/datatypes/test_temporal.py [hunk 11]
```diff
@@ -2744,3 +2744,148 @@ def test_rolling_duplicates() -> None:
...
+                        "date_col",
+                        [
+                            datetime(2020, 10, 10, 0, 0),
+                            datetime(2020, 10, 15, 0, 0),
+                            datetime(2020, 10, 15, 0, 0),
+                        ],
+                        dtype=pl.Datetime(time_unit="us", time_zone="UTC"),
+                    ).dt.convert_time_zone("Europe/London"),
+                ]
+            ),
```
Comment (MarcoGorelli) [inline py-polars/tests/unit/datatypes/test_temporal.py:2880]: this is a huge parametrisation 😄 

to clean it up, would you just parametrise over the values (`[datetime(2020, 10, 10), ...]`) and `schema`?

same for the other test

File: crates/polars-arrow/src/legacy/kernels/time.rs [hunk 12]
```diff
@@ -70,3 +70,39 @@ pub fn convert_to_naive_local_opt(
         LocalResult::None => None,
     }
 }
+
+#[cfg(feature = "timezones")]
+pub fn naive_utc_to_naive_local_in_new_time_zone(
```
Comment (MarcoGorelli) [inline crates/polars-arrow/src/legacy/kernels/time.rs:75]: nice descriptive names 👍 do you want to make a separate PR to make `convert_to_naive_local_opt` similarly descriptive?

**Review by MarcoGorelli — COMMENTED** [review]
this is coming along nicely!

File: py-polars/polars/expr/datetime.py (hunk 9 above)
Comment (ion-elgreco) [reply py-polars/polars/expr/datetime.py:1467]: But then if you want to pass a column, you have to do pl.col("tz"), right?

File: py-polars/polars/expr/datetime.py (hunk 10 above)
Comment (MarcoGorelli) [reply py-polars/polars/expr/datetime.py:1476]: `try_ternary_elementwise`

Comment (ion-elgreco) [reply py-polars/polars/expr/datetime.py:1476]: Nice, will try to build with that in the follow up :D

File: crates/polars-ops/src/chunked_array/datetime/replace_time_zone.rs [hunk 13]
```diff
@@ -82,3 +84,56 @@ pub fn replace_time_zone(
...
+            Some(convert_tz) => datetime.0.try_apply(|timestamp| {
+                let ndt = timestamp_to_datetime(timestamp);
+                Ok(datetime_to_timestamp(
+                    convert_to_new_timezone_and_naive_local(
+                        &from_tz,
+                        &parse_time_zone(convert_tz)?,
+                        ndt,
+                    )?,
+                ))
+            }),
```
[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply crates/polars-ops/src/chunked_array/datetime/replace_time_zone.rs]: `try_apply` takes a function which could throw an error, and returns a result

`apply` take a function which doesn't fail, and just returns a chunked array

They should both handle null values

`naive_utc_to_naive_local_in_new_time_zone` can't fail, so it should be fine to use `apply`

File: py-polars/polars/expr/datetime.py [hunk 14]
```diff
@@ -1423,6 +1423,113 @@ def replace_time_zone(
...
+            Time zone for the `Datetime` expression.
+
+        Returns
+        -------
+        Expr
+            Expression of data type :class:`DateTime`.
+
+        Examples
+        --------
+        >>> from datetime import datetime
```
[the remark this answers is not in the document]
Comment (ion-elgreco) [reply py-polars/polars/expr/datetime.py:1445]: Hmm, how about this:

to_local_datetime: "You can use `to_local_datetime` to figure out how a tz-aware datetime will be expressed in a local datetime given a timezone."

from_local_datetime: "You can go from a localized datetime back to expressing the datetimes in a single timezone with `from_local_datetime`."

File: crates/polars-ops/src/chunked_array/datetime/replace_time_zone.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (ion-elgreco) [reply crates/polars-ops/src/chunked_array/datetime/replace_time_zone.rs]: `naive_utc_to_naive_local_in_new_time_zone` indeed can't fail. But the timezone parsing with `parse_time_zone` could and this is happening in the scope of try_apply and of try_binary_elementwise, so I don't think we can change it to use apply.

File: py-polars/polars/expr/datetime.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply py-polars/polars/expr/datetime.py:1445]: I'd say "as a local datetime", but the rest looks good, nice!

File: crates/polars-ops/src/chunked_array/datetime/replace_time_zone.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply crates/polars-ops/src/chunked_array/datetime/replace_time_zone.rs]: there are two cases:
- `convert_tz` is length 1
- `convert_tz` is length n

in the first case, I think you can first parse the time zone, and then use `apply` (once you've parsed the time zone, the function can't fail?)

[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply crates/polars-ops/src/chunked_array/datetime/replace_time_zone.rs]: something like
```
Some(convert_tz) => {
    let to_tz = &parse_time_zone(convert_tz)?;
    datetime.0.apply(...
}
```

File: crates/polars-ops/src/chunked_array/datetime/replace_time_zone.rs [hunk 15]
```diff
@@ -82,3 +85,101 @@ pub fn replace_time_zone(
...
+    let datetime_to_timestamp: fn(NaiveDateTime) -> i64 = match datetime.time_unit() {
+        TimeUnit::Milliseconds => datetime_to_timestamp_ms,
+        TimeUnit::Microseconds => datetime_to_timestamp_us,
+        TimeUnit::Nanoseconds => datetime_to_timestamp_ns,
+    };
+    let out = match naive_tz.len() {
+        1 => match unsafe { naive_tz.get_unchecked(0) } {
+            Some(naive_tz) => datetime.0.try_apply(|timestamp| {
+                let ndt = timestamp_to_datetime(timestamp);
+                let from_tz = parse_time_zone(naive_tz)?;
```
Comment (MarcoGorelli) [inline crates/polars-ops/src/chunked_array/datetime/replace_time_zone.rs:162]: does this need to be parsed for each element? If it's just a single one, I'd suggest putting this one outside the `try_apply` as well

**Review by MarcoGorelli — COMMENTED** [review]
Awesome stuff!

Just got a minor comment, but I think the rest should be ready

File: py-polars/polars/expr/datetime.py [hunk 16]
```diff
@@ -1423,6 +1423,119 @@ def replace_time_zone(
             self._pyexpr.dt_replace_time_zone(time_zone, ambiguous._pyexpr)
         )
 
+    def to_local_datetime(
+        self,
+        tz: str | Expr,
+    ) -> Expr:
```
Comment (MarcoGorelli) [inline py-polars/polars/expr/datetime.py:1429]: should this method not be available for `polars.Series` as well?

File: py-polars/src/expr/datetime.rs [hunk 17]
```diff
@@ -50,6 +50,26 @@ impl PyExpr {
...
+        self.inner
+            .clone()
+            .dt()
+            .to_local_datetime(convert_tz.inner)
+            .into()
+    }
+
+    #[cfg(feature = "timezones")]
+    #[pyo3(signature = (naive_tz, out_tz, ambiguous))]
+    fn dt_from_local_datetime(&self, naive_tz: Self, out_tz: String, ambiguous: String) -> Self {
```
Comment (MarcoGorelli) [inline py-polars/src/expr/datetime.rs:65]: can we make the argument names match those from the python function?

File: py-polars/src/expr/datetime.rs [hunk 18]
```diff
@@ -50,6 +50,26 @@ impl PyExpr {
             .into()
     }
 
+    #[cfg(feature = "timezones")]
+    #[pyo3(signature = (convert_tz))]
+    fn dt_to_local_datetime(&self, convert_tz: Self) -> Self {
```
Comment (MarcoGorelli) [inline py-polars/src/expr/datetime.rs:55]: mismatched name too 😉

File: py-polars/polars/expr/datetime.py [hunk 19]
```diff
@@ -1423,6 +1423,119 @@ def replace_time_zone(
...
+        │ datetime[μs, UTC]       ┆ str              ┆ datetime[μs]        │
+        ╞═════════════════════════╪══════════════════╪═════════════════════╡
+        │ 2020-10-10 00:00:00 UTC ┆ Europe/London    ┆ 2020-10-10 01:00:00 │
+        │ 2020-10-10 00:00:00 UTC ┆ Africa/Kigali    ┆ 2020-10-10 02:00:00 │
+        │ 2020-10-10 00:00:00 UTC ┆ America/New_York ┆ 2020-10-09 20:00:00 │
+        └─────────────────────────┴──────────────────┴─────────────────────┘
+
+        """
+        local_tz = parse_as_expression(tz, str_as_lit=True)
+        return wrap_expr(self._pyexpr.dt_to_local_datetime(local_tz))
```
Comment (MarcoGorelli) [inline py-polars/polars/expr/datetime.py:1473]:
```suggestion
        tz = parse_as_expression(tz, str_as_lit=True)
        return wrap_expr(self._pyexpr.dt_to_local_datetime(tz))
```

File: py-polars/polars/expr/datetime.py [hunk 20]
```diff
@@ -1423,6 +1423,119 @@ def replace_time_zone(
...
+        ╞═════════════════════╪══════════════════╪═════════════════════════╡
+        │ 2020-10-10 01:00:00 ┆ Europe/London    ┆ 2020-10-10 00:00:00 UTC │
+        │ 2020-10-10 02:00:00 ┆ Africa/Kigali    ┆ 2020-10-10 00:00:00 UTC │
+        │ 2020-10-09 20:00:00 ┆ America/New_York ┆ 2020-10-10 00:00:00 UTC │
+        └─────────────────────┴──────────────────┴─────────────────────────┘
+        """
+        naive_time_zone = parse_as_expression(from_tz, str_as_lit=True)
+        return wrap_expr(
+            self._pyexpr.dt_from_local_datetime(naive_time_zone, out_tz, ambiguous)
+        )
```
Comment (MarcoGorelli) [inline py-polars/polars/expr/datetime.py:1537]:
```suggestion
        from_tz= parse_as_expression(from_tz, str_as_lit=True)
        return wrap_expr(
            self._pyexpr.dt_from_local_datetime(from_tz, out_tz, ambiguous)
        )
```

**Review by MarcoGorelli — COMMENTED** [review]
minor nitpicks (so sorry) - will sleep on it and approve tomorrow!

File: py-polars/src/expr/datetime.rs (hunk 18 above)
Comment (ion-elgreco) [reply py-polars/src/expr/datetime.rs:55]: Ah yep, will resolve that later

**Comment by ion-elgreco** [comment]
@ritchie46 @stinodego could someone take a look at this PR for second approval? Marco already reviewed and approved it 😄

**Review by stinodego — CHANGES_REQUESTED** [review]
I find `to_local_datetime` and `from_local_datetime` to be very confusing names. My local time zone has nothing to do with it. In fact, this is about datetimes without any time zone at all. "naive" is a more fitting term when talking about such datetimes, in my opinion.

I can see how the functions are useful though, but I think we need a bit more bikeshedding about the naming here.

**Comment by ion-elgreco** [comment]
@stinodego maybe these variations are better:



`to_localized_datetime` and `from_localized_datetime`

Or

 `to_naive_local_datetime` and `from_naive_local_datetime`

File: py-polars/polars/expr/datetime.py [hunk 21]
```diff
@@ -1423,6 +1423,114 @@ def replace_time_zone(
...
+        └─────────────────────────┴──────────────────┴─────────────────────┘
+
+        """
+        tz = parse_as_expression(tz, str_as_lit=True)
+        return wrap_expr(self._pyexpr.dt_to_local_datetime(tz))
+
+    def from_local_datetime(
+        self,
+        from_tz: str | Expr,
+        to_tz: str,
```
[the remark this answers is not in the document]
Comment (ion-elgreco) [reply py-polars/polars/expr/datetime.py:1475]: @stinodego from_time_zone to make it consistent with the other suggestion? Or naive_time_zone?

**Comment by ion-elgreco** [comment]
@stinodego can you make a final decision on the naming, I am happy to move forward with whatever you think is best, I want to start using the functionality :D

**Comment by ion-elgreco** [comment]
@stinodego closing it since, we will move it to polars-xdt : ) 

For reference: https://github.com/MarcoGorelli/polars-xdt/pull/42
